### PR TITLE
Case 7295: Fix rendering of unrigged vertices

### PR DESCRIPTION
**Pull request by SamGondelman**
https://highfidelity.manuscript.com/f/cases/7295/Unrigged-Vertices-On-Rigged-Model-Go-Into-Active-User-s-Eye-Points

Test plan:
- See the steps here: https://highfidelity.manuscript.com/f/cases/7295/Unrigged-Vertices-On-Rigged-Model-Go-Into-Active-User-s-Eye-Points#BugEvent.146316
- When the model appears, it won't have a hole in its chest.
- When it is scaled non-uniformly, the chest vertices will still render normally.

### CODE REVIEW

File: libraries/fbx/src/FBXSerializer.cpp [hunk 1]
```diff
@@ -1510,19 +1512,6 @@ HFMModel* FBXSerializer::extractHFMModel(const QVariantHash& mapping, const QStr
...
-                glm::vec3 boneDirection;
-                float boneLength = 0.0f;
-                if (joint.parentIndex != -1) {
-                    boneBegin = extractTranslation(inverseModelTransform * hfmModel.joints[joint.parentIndex].bindTransform);
-                    boneDirection = boneEnd - boneBegin;
-                    boneLength = glm::length(boneDirection);
-                    if (boneLength > EPSILON) {
-                        boneDirection /= boneLength;
-                    }
-                }
```
Comment (SamGondelman) [inline libraries/fbx/src/FBXSerializer.cpp]: These were just unused

File: libraries/fbx/src/FBXSerializer.cpp [hunk 2]
```diff
@@ -1482,8 +1486,8 @@ HFMModel* FBXSerializer::extractHFMModel(const QVariantHash& mapping, const QStr
             }
         }
 
-        // if we don't have a skinned joint, parent to the model itself
-        if (extracted.mesh.clusters.isEmpty()) {
```
Comment (SamGondelman) [inline libraries/fbx/src/FBXSerializer.cpp]: Now we always add this to the end of the list, and fill `extracted.mesh.clusterIndices` with this index instead, so unrigged vertices default to the root transform.

File: libraries/fbx/src/FBXSerializer.cpp [hunk 3]
```diff
@@ -1575,20 +1565,24 @@ HFMModel* FBXSerializer::extractHFMModel(const QVariantHash& mapping, const QStr
...
+
+                const float ALMOST_HALF = 0.499f;
                 if (totalWeight > 0.0f) {
-                    const float ALMOST_HALF = 0.499f;
                     float weightScalingFactor = (float)(UINT16_MAX) / totalWeight;
                     for (int k = j; k < j + WEIGHTS_PER_VERTEX; ++k) {
                         extracted.mesh.clusterWeights[k] = (uint16_t)(weightScalingFactor * weightAccumulators[k] + ALMOST_HALF);
                     }
+                } else {
+                    extracted.mesh.clusterWeights[j] = (uint16_t)((float)(UINT16_MAX) + ALMOST_HALF);
```
Comment (SamGondelman) [inline libraries/fbx/src/FBXSerializer.cpp]: If this vertex didn't get any weights, give it a weight of 1 here

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/18059/HighFidelity-Beta-PR15158-b24b7fe.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/18059/HighFidelity-Beta-Interface-PR15158-b24b7fe.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/18059/HighFidelity-Beta-PR15158-b24b7fe.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/18059/HighFidelity-Beta-Interface-PR15158-b24b7fe.exe

**Comment by SamGondelman** [comment]
build this please

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/18065/HighFidelity-Beta-PR15158-b24b7fe.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/18065/HighFidelity-Beta-Interface-PR15158-b24b7fe.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/18065/HighFidelity-Beta-PR15158-b24b7fe.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/18065/HighFidelity-Beta-Interface-PR15158-b24b7fe.exe

**Comment by SamGondelman** [comment]
build this please

File: libraries/render-utils/src/Model.cpp [hunk 4]
```diff
@@ -247,15 +247,15 @@ void Model::updateRenderItems() {
                 Transform renderTransform = modelTransform;
 
                 if (useDualQuaternionSkinning) {
-                    if (meshState.clusterDualQuaternions.size() == 1) {
+                    if (meshState.clusterDualQuaternions.size() <= 2) {
                         const auto& dq = meshState.clusterDualQuaternions[0];
```
Comment (sabrina-shanman) [inline libraries/render-utils/src/Model.cpp]: Do we have to worry about the case when clusterDualQuaternions.size() == 0?

Comment (SamGondelman) [reply libraries/render-utils/src/Model.cpp]: changed it just in case 👍

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/18070/HighFidelity-Beta-PR15158-2b32b77.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/18070/HighFidelity-Beta-Interface-PR15158-2b32b77.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/18070/HighFidelity-Beta-PR15158-2b32b77.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/18070/HighFidelity-Beta-Interface-PR15158-2b32b77.exe

**Comment by SamGondelman** [comment]
build this please

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/18080/HighFidelity-Beta-PR15158-2b32b77.dmg

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/18080/HighFidelity-Beta-PR15158-2b32b77.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/18080/HighFidelity-Quest-Beta-PR15158-2b32b77.apk)

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/18080/HighFidelity-Beta-PR15158-2b32b77.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/18080/HighFidelity-Beta-Interface-PR15158-2b32b77.exe

**Comment by Atlante45** [comment]
@SamGondelman did you mean to commit `.vs/slnx.sqlite`?
I just noticed it in master.

**Comment by huffman** [comment]
I just noticed this to. Can we get it removed and added to `.gitignore`?

**Pull request merged**